### DISCUSSION TITLE
[Proposal]: override constructor inheritance

### DISCUSSION THREAD

**Discussion by Thaina** [dotnet/csharplang#4496] [Language Ideas] [answered]
* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary

I want to have an simple way to declare that, I want this class to always inherit every missing constructor of the base class by default

```C#
public abstract class A
{
    public A(int x,int y) { }
    public A(float x,float y) { }
}

public abstract class B : A
{
    override base; // a syntax to inherit all constructor
}

public class C : B
{
    public C() : base(0,0) { } // direct constructing to A()
}
```

It actually transpile at compile time

```C#
public abstract class A
{
    public A(int x,int y) { }
    public A(float x,float y) { }
}

public abstract class B : A
{
    public B(int x,int y) : base(x,y){ }
    public B(float x,float y) : base(x,y) { }
}

public class C : B
{
    public C() : base(0,0) { } // direct constructing to A(int,int)
}
```

## Motivation

Sometimes we might have complex inheritance class that can group into some set of classes. And there are some chance that along the hierarchy we might have bloated constructor in base class and the separate group of classes that don't require constructor

Sorry that was hard to explain, but let me make some example. Suppose there is 4 classes related to UI

```C#
class Panel
{
    public Rect rect;
    public void UpdateUI() { }

    public SomeOverlapField0 o0;
    public void PanelFunction() { }

    public SomeUniqueField0 u0;
}

class Image
{
    public Rect rect;
    public void UpdateUI() { }

    public SomeOverlapField0 o0;
    public void PanelFunction() { }

    public SomeUniqueField1 u1;
}

class Table
{
    public Rect rect;
    public void UpdateUI() { }

    public SomeOverlapField1 o1;
    public void TableFunction() { }

    public SomeUniqueField2 u2;
}

class TableRow
{
    public Rect rect;
    public void UpdateUI() { }

    public SomeOverlapField1 o1;
    public void TableFunction() { }

    public SomeUniqueField3 u3;
}
```

I could make an abstract base class for every overlapped things like to

```C#
abstract class UI
{
    public Rect rect;
    public abstract void UpdateUI();
}

abstract class PanelBase : UI
{
    public SomeOverlapField0 o0;
    public abstract void PanelFunction();
}

abstract class TableBase : UI
{
    public SomeOverlapField1 o1;
    public abstract void TableFunction() { }
}

class Panel : PanelBase
{
    public SomeUniqueField0 u0;
    public override void UpdateUI() { }
    public override void PanelFunction() { }
}

class Image : PanelBase
{
    public SomeUniqueField1 u1;
    public override void UpdateUI() { }
    public override void PanelFunction() { }
}

class Table : TableBase
{
    public SomeUniqueField2 u2;
    public override void UpdateUI() { };
    public override void TableFunction() { }
}

class TableRow : TableBase
{
    public SomeUniqueField3 u3;
    public override void UpdateUI() { };
    public override void TableFunction() { }
}
```

Seem fine and all. But the problem would came whenever we want to make constructor for the base class. We then need to make various constructor, especially in `PanelBase` and `TableBase` that might not need to do anything except mapping exact constructor of the same parameter as a middle man. Instead of letting derived class refer to the top base class directly

## Detailed design

For simplicity, I think we should have

```C#
abstract class B : A
{
   override base;
}
```

Just this line, in any abstract derived class, to mark that this class will copy every constructors of the base class

## Drawbacks

The syntax I could think of is somehow weird. I don't know any better syntax but it could be anything better

## Alternatives

We always need to tediously copy and mimic constructor of base class. And change it everytime

We may write custom code generator. But it seem unweildy and need to be attach to every project we want. It also inject code into our codebase while if we have this language feature it would always hidden from sight by only one line of code

## Unresolved questions

1. I don't know should we allow this feature to every class or should we limit to only abstract class. For instanceable class it would allow `new` with base parameter. Which might be convenient, but I'm not sure there would be any conflict
2. Should we also allow normal constructor or should we limit that, if we use this feature then this class must not have any normal constructor? And if we allow it, would we allow constructor with the same signature? This feature then only copy the missing constructors to derived class

**Comment by YairHalberstadt** [comment]
I think source generators could be a good solution to this. You could have an attribute `GenerateBaseConstructors` which a source generator looks for and generates the desired constructors for.

**Reply by MichaRotstein** [reply]
Do you know of any open source implementation of such a generator ?

**Comment by svick** [comment]
> We may write custom code generator. But it seem unweildy and need to be attach to every project we want. It also inject code into our codebase while if we have this language feature it would always hidden from sight by only one line of code

That sounds like an argument against all source generators. What makes this proposal special that source generator would not be appropriate for it? Do you think this is something that would be useful very often? (And if you do, do you have some evidence for that?)

**Comment by Thaina** [comment]
> > We may write custom code generator. But it seem unweildy and need to be attach to every project we want. It also inject code into our codebase while if we have this language feature it would always hidden from sight by only one line of code
> 
> That sounds like an argument against all source generators. What makes this proposal special that source generator would not be appropriate for it? Do you think this is something that would be useful very often? (And if you do, do you have some evidence for that?)

Yes, I think this pattern is common enough to be language feature. Everytime there is inheritance for more than 2 layers, or adding a middle layer between existing classes for grouping things. Constructor become hindrance and you need to make constructor for every class inbetween. Not to mention sometimes you want to change constructor, even just switch the place of parameter, the whole chain then become a mess

In my opinion, source generator is for specific feature/purpose that unique and special. But class hierachy can be anywhere in the coding phase of project

**Comment by HaloFour** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/806

**Comment by Thaina** [comment]
Thanks, still I think this should be issue
